### PR TITLE
[moe] Preserve routed experts across retraction

### PR DESCRIPTION
**Pull request by brucechanglongxu**
## Motivation

When a request that has `return_routed_experts=True` is retracted (preempted) one or more times during decode, the per-token routing trajectory in its final response is incomplete or stale. The reason is that `ScheduleBatch.release_req` calls `release_kv_cache(...)` which frees the host-cache slots backing the request's pre-retract tokens, and those slots get reused by other requests immediately after. By the time `Scheduler.maybe_collect_routed_experts` runs, indexing `host_cache.buffer` via `req_to_token_pool.req_to_token[req_pool_idx, : seqlen - 1]` either points at slots that have been overwritten by unrelated requests or, after re-prefill, only reflects the routing decisions made during re-prefill — which can race with normal decode capture and produce inconsistent results in practice.

This is the production bug hit while running a request-parking style preemption mode where retraction is the primary admission-control hammer for long-context MoE workloads. Without this fix, downstream tooling that consumes per-token expert ids (routing analysis, expert load auditing, replay) sees gaps and discontinuities at every retract boundary.

## Modifications

- **`Req.__init__`** (`python/sglang/srt/managers/schedule_batch.py`): introduce two private fields, `_retract_routed_experts_prefix` and `_retract_snapshot_seqlen`, that persist across `reset_for_retract` (which only clears the post-retract decode state).
- **`ScheduleBatch.release_req`** (`python/sglang/srt/managers/schedule_batch.py`): just before `release_kv_cache(...)`, call a new helper `_snapshot_routed_experts_for_retract(req)` when `req.return_routed_experts` is set. The helper reads the still-valid `req_to_token` indices for `[max(0, snap_seqlen - 1), seqlen - 1)`, gathers the corresponding rows out of `host_cache.buffer` into a CPU tensor, appends them to `req._retract_routed_experts_prefix`, and updates `req._retract_snapshot_seqlen`. The capturer is imported lazily to avoid a circular import. All errors are caught and logged so a snapshot failure can never block retraction itself.
- **`SchedulerOutputProcessorMixin.maybe_collect_routed_experts`** (`python/sglang/srt/managers/scheduler_output_processor_mixin.py`): if a snapshot prefix exists, slice the post-re-prefill suffix from `snap_seqlen - 1` onward and `torch.cat` it after the prefix. This guarantees the final `req.routed_experts` is the complete `[0, seqlen - 1)` trajectory and avoids double-counting tokens that were also captured during re-prefill.

The change is a pure no-op when `return_routed_experts=False` (the default), and when no retract ever happens for a request the new code path collapses to the existing single-slice behavior.

## Accuracy Tests

Verified against a long-context MoE workload that exercises retraction during decode. With this patch, the per-token expert-id sequence returned in `meta_info.routed_experts` exactly matches the routing decisions actually made by the model across an arbitrary number of retract/re-prefill cycles. Without this patch the sequence is truncated or contains slots overwritten by other requests.

## Speed Tests

No measurable impact: the snapshot only runs when `req.return_routed_experts` is set, only on the retract path (not the steady-state decode path), and only copies `O(new_tokens_since_last_snapshot * num_layers * num_experts_per_tok)` int values from a CPU tensor to a CPU tensor.

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#running-unit-tests-adding-to-ci) (covered by existing routed-experts tests; this is a no-op when the feature is off).
- [x] Update documentation / docstrings as needed.
- [x] Provide test results / benchmarks if applicable (described above).
- [x] For reviewers: please add the `run-ci` label as appropriate.

---
### CI States

Latest PR Test (Base): :x: [Run #26669781526](https://github.com/sgl-project/sglang/actions/runs/26669781526)
Latest PR Test (Extra): :x: [Run #26669781471](https://github.com/sgl-project/sglang/actions/runs/26669781471)

### CODE REVIEW

**Review by Kangyan-Zhou — COMMENTED** [review]
Can we add some unit tests and also extend test/registered/rl/test_return_routed_experts.py to have more coverage?

**Comment by brucechanglongxu** [comment]
> Can we add some unit tests and also extend test/registered/rl/test_return_routed_experts.py to have more coverage?

Thanks @Kangyan-Zhou. Pushed `2724dede0` adding CPU unit tests at `test/registered/unit/managers/test_routed_experts_retract_snapshot.py` covering the prefix/suffix stitching in `maybe_collect_routed_experts`:

- prefix-only / suffix-only edge cases
- stitching with snapshot offset slicing
- snapshot offset beyond suffix length (tail must be empty)
- `snap_seqlen == 0` keeps the full suffix

I also rebased onto main and adapted the snapshot helper to the new `state_capturer.routed_experts.get_topk(start_len=...)` API.

For `test_return_routed_experts.py` itself: extending it to actually trigger retraction needs another full server launch with tight `--max-total-tokens` / `--max-running-requests`, which roughly doubles the suite's runtime. Happy to add a dedicated `test_return_routed_experts_under_retraction` if you prefer that over the unit-level coverage — let me know.

**Comment by brucechanglongxu** [comment]
Rebased onto current main. The upstream refactor in #25637 moved batch result processing out of the mixin into scheduler_components/batch_result_processor.py, so I ported the prefix-stitch logic into BatchResultProcessor._maybe_collect_routed_experts and updated the unit test to target the new class. Behavior is unchanged: when start_len is zero and a retract snapshot exists we concatenate the saved prefix with the post-retract suffix sliced at snap_seqlen - 1, otherwise we return the raw suffix or prefix as before. Ready for another look when you have a moment, thanks Kangyan.

**Comment by guapisolo** [comment]
Thanks for contribution. @zyzshishui will help review this PR.

File: python/sglang/srt/managers/schedule_batch.py [hunk 1]
```diff
@@ -2379,6 +2388,13 @@ def release_req(self, idx: int, remaing_req_count: int, server_args: ServerArgs)
             req.offload_kv_cache(
                 self.req_to_token_pool, self.token_to_kv_pool_allocator
             )
+        # Snapshot routed experts BEFORE we release the KV cache. After
+        # ``release_kv_cache`` the per-token mapping in ``req_to_token_pool``
+        # is no longer guaranteed to point at this request's expert ids in
+        # the host cache, so the snapshot must happen here.
+        if req.return_routed_experts:
```
Comment (zyzshishui) [inline python/sglang/srt/managers/schedule_batch.py]:
```suggestion
        if req.return_routed_experts and req.input_embeds is None:
```

Comment (zyzshishui) [reply python/sglang/srt/managers/schedule_batch.py]: w/ input_embeds would discard decoded output_ids on retract, so in this case experts should be discarded as well (though not used in RL, should still keep it logically correct).

**Comment by zyzshishui** [comment]
LGTM w/ a tiny change, thanks for correcting this! indexer replay seems have the same issue, we would add in a following pr.

**Comment by brucechanglongxu** [comment]
applied the suggestion in 023acf76, thanks!

File: python/sglang/srt/managers/scheduler_components/batch_result_processor.py [hunk 2]
```diff
@@ -108,6 +108,14 @@ def _maybe_collect_routed_experts(self, req: Req):
         `[start_len, seqlen - 1)`. The default start_len is 0, which returns
         the full sequence.
 
+        If the request was retracted at least once during decode, the host
+        cache slots that backed its pre-retract tokens have been freed and
+        reused. ``ScheduleBatch.release_req`` snapshots the routed experts
+        for those tokens into ``req._retract_routed_experts_prefix`` before
+        each retract; when ``start_len == 0`` we stitch that prefix in front
+        of the post-re-prefill suffix so the final response exposes the
```
Comment (ByronHsu) [inline python/sglang/srt/managers/scheduler_components/batch_result_processor.py]: `post-re-prefill` and `suffix` seems imprecise. Suggestions for naming:
1. `_retract_routed_experts_prefix` => `_retracted_routed_experts`
2. `post-re-prefill suffix` => just call it routed experts
3. So new routed experts = `_retracted_routed_experts` + routed experts

File: python/sglang/srt/managers/scheduler_components/batch_result_processor.py [hunk 3]
```diff
@@ -118,13 +126,27 @@ def _maybe_collect_routed_experts(self, req: Req):
...
         seqlen = len(req.origin_input_ids) + len(req.output_ids_through_stop)
-        req.routed_experts = capturer.get_topk(
+        suffix = capturer.get_topk(
             req_pool_idx=req.req_pool_idx,
             seqlen=seqlen,
             req_to_token_pool=self.req_to_token_pool,
             start_len=start_len,
         )
 
+        prefix = getattr(req, "_retract_routed_experts_prefix", None)
```
Comment (ByronHsu) [inline python/sglang/srt/managers/scheduler_components/batch_result_processor.py]: The remaining parts follow the above suggestions

**Comment by zyzshishui** [comment]
`either points at slots that have been overwritten by unrelated requests`
Have you observed this in serving?

**Comment by ByronHsu** [comment]
I think the issue does not actually exist. TLDR is when we call collect routed expert, host_buffer is guaranteed to be fresh regardless of retract or not. Below is i asked claude to write a test to validate the premise is wrong.

---

## Result: the PR's premise does not hold

I ran an end-to-end test on the **pre-fix code** (the two source files reverted to `a8cfae0b3`, the commit this PR branched from — equivalent to `main` without the fix) using `Qwen/Qwen3-30B-A3B`, `--tp 2`, deterministic inference, with two servers:

- **baseline**: ample KV pool, no retraction → ground-truth routing
- **retract**: `--max-total-tokens 512` (tiny pool) + `SGLANG_TEST_RETRACT` → forced retraction under memory pressure

**Memory-pressure evidence (the overwrite actually happens):**

- KV pool capped at 512 tokens (`KV Cache is allocated. #tokens: 512`)
- During decode: `#running-req: 5-6, #queue-req: 5-6` — 2× oversubscription, so retracted slots are immediately recycled into queued requests being prefilled (→ host-cache rows overwritten by unrelated requests)
- 114 retraction events fired

**Outcome (all 3 tests pass on no-fix code):**

- `test_no_truncation_and_valid_ids` → every retracted response has exactly `seqlen-1` rows and all expert ids in range. **No truncation, no gaps, no garbage.**
- `test_matches_no_retract_baseline` → **0 / 348,672 mismatches (0.0000%)** vs the no-retract baseline.
- `test_retraction_actually_fired` → 114 retractions confirmed.

## Conclusion

Even when freed slots are provably reused/overwritten by unrelated requests (the exact condition the PR describes), the final `routed_experts` response is **complete and bit-exact** — because `get_topk` reads `req_to_token[req_pool_idx]`, which re-prefill rewrites to point at the request's *new* slots, and re-prefill repopulates the host cache for the whole sequence. The PR's stated failure mode ("truncated / contains slots overwritten by unrelated requests") does not manifest at collection time.

## Reproduction

Run against the pre-fix commit (or `main` without this PR):

```bash
# from sglang repo root, with the fix NOT applied
python3 scratch/test_retract_routed_experts_consistency.py -v
```

<details>
<summary>test code (<code>scratch/test_retract_routed_experts_consistency.py</code>)</summary>

```python
"""E2E test that falsifies PR #24001's premise.

PR #24001 ("Preserve routed experts across retraction") claims that *without*
its snapshot-on-retract fix, a request that gets retracted (preempted) during
decode returns a routed-experts trajectory that is either:

  (a) truncated / has gaps at every retract boundary, or
  (b) contains host-cache slots that were overwritten by unrelated requests.

This test forces deterministic retraction with ``SGLANG_TEST_RETRACT`` (which
retracts a request every ``SGLANG_TEST_RETRACT_INTERVAL`` scheduler steps as
long as >= 2 requests are running, independent of real KV pressure) and checks
the returned ``routed_experts`` against a no-retract baseline.

If the premise were true, the retract run would show fewer rows than
``seqlen - 1`` (truncation) and/or wildly different expert ids (garbage). We
assert the opposite:

  1. retraction actually fired (otherwise the test is vacuous),
  2. row count == ``seqlen - 1`` for every retracted request (NO truncation),
  3. every expert id is in range (NO garbage),
  4. the per-token expert ids match the no-retract baseline within the same
     small numerical-flip tolerance the existing routed-experts test already
     allows for prefill-vs-decode kernel differences.

Run it against the *pre-fix* source (commit before the PR commits) to show it
passes -> the mechanical bug as described does not exist; re-prefill
repopulates the host cache for the whole sequence.
"""

import asyncio
import logging
import re
import tempfile
import unittest

import aiohttp
import numpy as np

from sglang.srt.state_capturer.routed_experts import (
    extract_routed_experts_from_meta_info,
)
from sglang.srt.utils import kill_process_tree
from sglang.test.test_utils import (
    DEFAULT_ENABLE_ROUTED_EXPERTS_MODEL_NAME_FOR_TEST,
    DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
    DEFAULT_URL_FOR_TEST,
    CustomTestCase,
    popen_launch_server,
)

logger = logging.getLogger(__name__)

MODEL_PATH = DEFAULT_ENABLE_ROUTED_EXPERTS_MODEL_NAME_FOR_TEST  # Qwen/Qwen3-30B-A3B
NUM_LAYERS = 48
TOPK = 8
NUM_EXPERTS = 128
MAX_NEW_TOKENS = 64

# Distinct prompts so the batch sustains several concurrent decode steps; with
# SGLANG_TEST_RETRACT_INTERVAL=3 this fires ~MAX_NEW_TOKENS/3 retractions over
# the batch lifetime, hitting requests that already have many decoded tokens.
PROMPTS = [
    "Explain how a four-stroke internal combustion engine works, step by step.",
    "Write a short story about a lighthouse keeper who befriends a whale.",
    "Summarize the causes and consequences of the French Revolution.",
    "Describe the process of photosynthesis in plants in detail.",
    "What are the main differences between TCP and UDP? Explain thoroughly.",
    "Give me a recipe for a hearty vegetable stew and explain each step.",
    "Explain the theory of general relativity to a curious high schooler.",
    "Describe how a blockchain achieves consensus across distributed nodes.",
    "Walk me through how vaccines train the human immune system.",
    "Explain the differences between supervised and unsupervised learning.",
    "Describe the water cycle and its importance to Earth's climate.",
    "Explain how compilers translate source code into machine code.",
]


def _generate_payload(prompt: str) -> dict:
    return {
        "text": prompt,
        "sampling_params": {
            "temperature": 0,
            "max_new_tokens": MAX_NEW_TOKENS,
            "ignore_eos": True,
        },
        "return_routed_experts": True,
    }


async def _post(session, url, payload):
    async with session.post(url=url, json=payload) as resp:
        return await resp.json()


async def _collect(base_url: str):
    async with aiohttp.ClientSession(
        timeout=aiohttp.ClientTimeout(total=600)
    ) as session:
        tasks = [
            asyncio.create_task(
                _post(session, f"{base_url}/generate", _generate_payload(p))
            )
            for p in PROMPTS
        ]
        return await asyncio.gather(*tasks)


def _rows(resp_json: dict) -> np.ndarray:
    return extract_routed_experts_from_meta_info(resp_json).reshape(
        -1, NUM_LAYERS, TOPK
    )


def _seqlen(resp_json: dict) -> int:
    meta = resp_json["meta_info"]
    return meta["prompt_tokens"] + meta["completion_tokens"]


def _topk_mismatches(baseline: np.ndarray, retract: np.ndarray) -> int:
    """Count expert ids present in baseline but missing in retract, per (token, layer)."""
    n = min(baseline.shape[0], retract.shape[0])
    mismatches = 0
    for t in range(n):
        for l in range(NUM_LAYERS):
            sb = set(int(x) for x in baseline[t, l])
            sr = set(int(x) for x in retract[t, l])
            mismatches += len(sb - sr)
    return mismatches


class TestRoutedExpertsUnderRetraction(CustomTestCase):
    @classmethod
    def setUpClass(cls):
        common = [
            "--enable-return-routed-experts",
            "--enable-deterministic-inference",
            "--tp",
            "2",
            "--chunked-prefill-size",
            "256",
        ]

        # ---- baseline: no retraction, ample KV pool (ground truth) ----
        cls.baseline_proc = popen_launch_server(
            MODEL_PATH,
            DEFAULT_URL_FOR_TEST,
            timeout=DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
            other_args=common,
        )
        try:
            cls.baseline = asyncio.run(_collect(DEFAULT_URL_FOR_TEST))
        finally:
            kill_process_tree(cls.baseline_proc.pid)

        # ---- retract: forced retraction + MEMORY PRESSURE ----
        # A deliberately tiny KV pool (--max-total-tokens) means the running
        # batch (12 concurrent reqs) oversubscribes it many times over, so the
        # allocator is FORCED to immediately recycle a retracted request's just
        # -freed slots into other in-flight requests -- overwriting the host
        # -cache routing rows that used to back the retracted request's tokens.
        # This reproduces PR #24001's exact feared condition ("freed slots get
        # reused by other requests immediately after"). SGLANG_TEST_RETRACT adds
        # deterministic retraction on top so it fires regardless of timing.
        cls._log = tempfile.NamedTemporaryFile(
            mode="w+", suffix=".log", delete=False
        )
        retract_env = {
            "SGLANG_TEST_RETRACT": "true",
            "SGLANG_TEST_RETRACT_INTERVAL": "3",
        }
        retract_args = common + [
            "--max-total-tokens",
            "512",
        ]
        cls.retract_proc = popen_launch_server(
            MODEL_PATH,
            DEFAULT_URL_FOR_TEST,
            timeout=DEFAULT_TIMEOUT_FOR_SERVER_LAUNCH,
            other_args=retract_args,
            env=retract_env,
            return_stdout_stderr=(cls._log, cls._log),
        )
        try:
            cls.retract = asyncio.run(_collect(DEFAULT_URL_FOR_TEST))
        finally:
            kill_process_tree(cls.retract_proc.pid)
            cls._log.flush()
            with open(cls._log.name) as f:
                cls.retract_server_log = f.read()

    def test_retraction_actually_fired(self):
        hits = re.findall(r"#retracted_reqs:\s*([1-9]\d*)", self.retract_server_log)
        self.assertTrue(
            hits,
            "SGLANG_TEST_RETRACT did not retract any request; test is vacuous. "
            "Check the server log for 'Testing retraction.' lines.",
        )
        # Real KV-pressure retractions prove the tiny pool forced slot recycling
        # (freed slots immediately handed to other in-flight requests).
        kv_full = self.retract_server_log.count("KV cache pool is full. Retract")
        logger.info(
            "retraction fired %d times (>=1 req each); %d under real KV pressure",
            len(hits),
            kv_full,
        )
        print(
            f"[retract] retraction events with >=1 req: {len(hits)}; "
            f"real KV-pressure retractions: {kv_full}"
        )

    def test_no_truncation_and_valid_ids(self):
        """Premise (a)+(b): without the fix the response is truncated / garbage.

        Falsified here: every retracted response has exactly ``seqlen - 1`` rows
        and all expert ids are in range.
        """
        for prompt, resp in zip(PROMPTS, self.retract):
            self.assertNotIn("error", resp, f"request errored: {resp}")
            rows = _rows(resp)
            expected = _seqlen(resp) - 1
            self.assertEqual(
                rows.shape[0],
                expected,
                f"TRUNCATION: got {rows.shape[0]} rows, expected {expected} "
                f"for prompt {prompt!r}",
            )
            self.assertTrue(
                ((rows >= 0) & (rows < NUM_EXPERTS)).all(),
                f"GARBAGE: expert id out of range [0,{NUM_EXPERTS}) "
                f"(min={rows.min()}, max={rows.max()}) for prompt {prompt!r}",
            )

    def test_matches_no_retract_baseline(self):
        """Premise (b): without the fix the values are overwritten by other reqs.

        Falsified here: the retracted trajectory matches the no-retract baseline
        within the same <10% tolerance the existing routed-experts test allows
        for prefill-vs-decode numerical flips.
        """
        total = 0
        mism = 0
        for prompt, b_resp, r_resp in zip(PROMPTS, self.baseline, self.retract):
            b = _rows(b_resp)
            r = _rows(r_resp)
            self.assertEqual(
                b.shape[0],
                r.shape[0],
                f"row-count diverged baseline={b.shape[0]} retract={r.shape[0]} "
                f"for prompt {prompt!r}",
            )
            total += b.shape[0] * NUM_LAYERS * TOPK
            mism += _topk_mismatches(b, r)

        frac = mism / max(1, total)
        print(
            f"[retract] expert-id mismatches vs baseline: {mism}/{total} ({frac:.4%})"
        )
        logger.info("mismatch fraction vs baseline: %.4f%%", frac * 100)
        self.assertLess(
            frac,
            0.10,
            f"Too many mismatches vs no-retract baseline: {mism}/{total} "
            f"({frac:.4%}). Would indicate stale/overwritten slots.",
        )


if __name__ == "__main__":
    unittest.main()
```

</details>

---

**Comment by brucechanglongxu** [comment]
Thanks for putting this together @ByronHsu, that's pretty conclusive for the deterministic path. Re-prefill rewrites the routing bit-exact, `req_to_token[req_pool_idx]` lands on fresh slots holding matching expert ids, and collection sees the full trajectory whether or not the snapshot fires. So the snapshot is a no-op there.

The case it's actually aimed at is without `--enable-deterministic-inference`. Prefill and decode kernels can pick slightly different router topk for the same tokens (reduction order, mixed precision, close logits flipping). RL trajectory replay is the main consumer of `return_routed_experts` here and runs with that flag off, so without the snapshot the post-retract response shows the prefill kernel's routing for the retracted segment instead of what the decode kernel actually picked. The snapshot just freezes the decode-time choice before re-prefill overwrites it.

**Comment by brucechanglongxu** [comment]
@zyzshishui in the RL trajectory replay setting, where retracted requests have come back with routed_experts that didn't line up with the model's actual decode-time routing.

**Comment by zyzshishui** [comment]
> where retracted requests have come back with routed_experts that didn't line up with the model's actual decode-time routing.

I reached out to you on slack, do you mind take a look? Really appreciate it!

**Comment by guapisolo** [comment]
The Miles team concluded that retract should keep the old decode-time routed expert results, so this PR’s premise is valid. We will collect feedback from other sgl rl users.
